src/neo/SmartContract/Native/Tokens/GasToken.cs

```diff
@@ -35,15 +35,34 @@ internal override bool Initialize(ApplicationEngine engine)
             return true;
         }
 
+        /// <summary>
+        /// The paid final paid amount is calculated based based on the sysfee and sysfeeCredit
+        /// </summary>
+        /// <param name="tx"></param>
+        /// <returns></returns>
+        private long CalculateSystemFeeWithPayback(Transaction tx)
+        {
+            long gasUnit = (long)GAS.Factor;
+            long finalFee = Math.Max(tx.SystemFee + tx.SysFeeCredit, 0);
+            long remainder = finalFee % gasUnit;
+            finalFee += gasUnit - remainder;
+            return finalFee;
+        }
+
         protected override bool OnPersist(ApplicationEngine engine)
         {
             if (!base.OnPersist(engine)) return false;
+            long sysFeeWithStorageDiscount = 0;
             foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
-                Burn(engine, tx.Sender, tx.SystemFee + tx.NetworkFee);
+            {
+                var sysFee = CalculateSystemFeeWithPayback(tx);
```

Comment:
In `CalculateSystemFeeWithPayback`, `tx.SysFeeCredit` = 0 ? As `GasToken.OnPersist` executes before tx.script.

Reply:
I will add more tests, so far I was only able to ensure that the payback is being calculated correctly. The next step is to ensure that we are charging the user the adequate fee.